datastore/protocol.go

```diff
@@ -0,0 +1,281 @@
+	updateObjectPeriodByDataIDWithPasswordHandler       func(dataID uint64, dataType uint16, password uint64) uint32
+	updateObjectMetaBinaryByDataIDWithPasswordHandler   func(dataID uint64, metaBinary []byte, password uint64) uint32
+	updateObjectDataTypeByDataIDWithPasswordHandler     func(dataID uint64, period uint16, password uint64) uint32
+	getObjectSizeByDataIDHandler                        func(dataID uint64) (uint32, uint32)
+	updateObjectUploadCompletedByDataIDHandler          func(dataID uint64, uploadCompleted bool) uint32
+	getObjectInfoByPersistenceTargetWithPasswordHandler func(persistenceTarget *datastore_types.DataStorePersistenceTarget, password uint64) (*datastore_types.DataStoreMetaInfo, uint32)
+	getObjectInfoByDataIDWithPasswordHandler            func(dataID uint64, password uint64) (*datastore_types.DataStoreMetaInfo, uint32)
+	s3GetRequestHeadersHandler                          func() ([]*datastore_types.DataStoreKeyValue, uint32)
+	s3PostRequestHeadersHandler                         func() ([]*datastore_types.DataStoreKeyValue, uint32)
+	initializeObjectByPreparePostParamHandler           func(ownerPID uint32, param *datastore_types.DataStorePreparePostParam) (uint64, uint32)
+	initializeObjectRatingWithSlotHandler               func(dataID uint64, param *datastore_types.DataStoreRatingInitParamWithSlot) uint32
+	rateObjectWithPasswordHandler                       func(dataID uint64, slot uint8, ratingValue int32, accessPassword uint64) (*datastore_types.DataStoreRatingInfo, uint32)
+	deleteObjectByDataIDWithPasswordHandler             func(dataID uint64, password uint64) uint32
+	deleteObjectByDataIDHandler                         func(dataID uint64) uint32
+	getObjectInfosByDataStoreSearchParamHandler         func(param *datastore_types.DataStoreSearchParam) ([]*datastore_types.DataStoreMetaInfo, uint32, uint32)
+}
+
+func (c *CommonDataStoreProtocol) S3StatObject(bucket, key string) (minio.ObjectInfo, error) {
+	return c.MinIOClient.StatObject(context.TODO(), bucket, key, minio.StatObjectOptions{})
+}
+
+func (c *CommonDataStoreProtocol) S3ObjectSize(bucket, key string) (uint64, error) {
+	info, err := c.S3StatObject(bucket, key)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(info.Size), nil
+}
+
+func (c *CommonDataStoreProtocol) VerifyObjectPermission(ownerPID, accessorPID uint32, permission *datastore_types.DataStorePermission) uint32 {
+	if permission.Permission > 3 {
+		return nex.Errors.DataStore.InvalidArgument
+	}
+
+	// * Allow anyone
+	if permission.Permission == 0 {
+		return 0
+	}
+
+	// * Allow friends
+	if permission.Permission == 1 {
+		friendsList := c.getUserFriendPIDsHandler(ownerPID)
+
+		if !slices.Contains(friendsList, accessorPID) {
+			return nex.Errors.DataStore.PermissionDenied
+		}
+	}
+
+	// * Allow people in permission.RecipientIDs
+	if permission.Permission == 2 {
+		if !slices.Contains(permission.RecipientIDs, accessorPID) {
+			return nex.Errors.DataStore.PermissionDenied
+		}
```

Comment:
For what I've seen, the recipient ID list doesn't include the PID that creates the object. A separate check should be added